### ISSUE TITLE
ROWCOUNT limited ON SqlServer 

### ISSUE THREAD

**Issue by KidYoungChang** [metabase/metabase#9940]
System : Win 10，Java 1.8
Data On SQL Server 2008
The results is affected by ROWCOUNT limited when use "Native Query",if the Metabase version >0.31.0.

```
Test SQL script 
----------------------------------------
DECLARE @DATA AS TABLE(
    IDX INT IDENTITY(1,1),
    V INT
)
DECLARE @STEP INT 
SET @STEP = 1
WHILE @STEP <=3000
BEGIN
    INSERT INTO @DATA(V)
    SELECT 1
    SET @STEP = @STEP + 1
END 

DECLARE @TEMP AS TABLE(
    IDX INT IDENTITY(1,1),
    V INT
)
INSERT INTO @TEMP(V)
SELECT V FROM @DATA

SELECT COUNT(1) FROM @TEMP
---------------------------------------

```
The data has 3000 rows,maybe more,but the result is limited at 2000.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by flamber** [comment]
Just for reference, this was added to the driver in 0.31.1:
https://github.com/metabase/metabase/blob/b913f5f70d8103d80255ccfcc2e09ae502c08220/src/metabase/driver/sqlserver.clj#L159-L169
Discussed in the forum last month: https://discourse.metabase.com/t/when-update-to-0-32-1-with-data-view-problem-on-sql-server/5811

**Comment by mazameli** [comment]
@camsaul It looks like this was changed in 0.31.1 in order to fix something, but seems to be causing unexpected results for users. Bug?

**Comment by notrom** [comment]
Metabase Version: v0.33.0-preview2 (also main branch v0.32.4)
Metabase DB: Postgres
Database: SQL Server 2014

I've struck this a few times recently. Really frustrating when you get a relatively complex query running "fast enough" in SQL using some indexed temp tables etc., then you move it into Metabase and it misses most of your data.

No real issues with limiting the rowcount for the end result set, although it would be good to be able to see/set this, but it would be great if we were free to do intermediate steps with no limit.

**Comment by notrom** [comment]
I just read the linked discourse thread. Based on that I found that if you add an explicit `SET ROWCOUNT 1000000` statement at the top of the script, or before the large dataset is selected, then you can work around it.

This is totally sufficient for me now that I know, I just wasn't aware that it was doing an explicit `SET ROWCOUNT 2000` on my session.

**Comment by dfx413** [comment]
Hi @flamber, just want to clarify if you don't mind. I have looked into the code (I don't use Clojure so I might be wrong) and if I read it correctly the referenced part works like this:
if there is ordering applied in the subquery it adds "TOP [absolute-max-results]" to the select statement so in my case it looks like `SELECT TOP 1048576 [whatever] FROM [whatever]`.
AFAIK this does not have to do anything with `SET ROWCOUNT` because id does not affect the count of rows returned in subquery, just the count of rows returned from the top-level query. 
So I don't really see a problem here. Am I wrong?

**Comment by flamber** [comment]
@dfx413 I'm not a Clojure programmer.

**Comment by dfx413** [comment]
Could someone else comment on this, please? I am really confused by this right now. Maybe @sbelak would know more? Thank you very much.

**Comment by MarcGJA** [comment]
Ran into what sounds like the same issue but with SQL Server TVPs that act rather like temp tables:
<a href="https://discourse.metabase.com/t/table-valued-parameters-tvp-sql-server-expressions-restricted-by-metabase-row-limit/7524">Table Valued Parameters (TVP) SQL Server expressions restricted by Metabase row limit</a>

Working around it with views & CTEs so far but it really threw me when time series data started being cut-off the further back in time the data reached!

Metabase: v0.33.2
Host: Localhost:3000
Databases: H2 internal default, Server SQL added
Browser: Firefox 52.9.0
OS: Microsoft Windows [Version 6.0.6002]
Java: 1.8.0_45

**Comment by camsaul** [comment]
Returning millions of rows shouldn't destroy the Metabase server since we stream the rows directly to the REST API JSON response as they come in... but it would probably destroy your browser and put a lot of load on the database.

Based on how it seems like they're not going to fix https://github.com/microsoft/mssql-jdbc/issues/176 upstream I think we should stop doing `.setMaxRows` for SQL Server and instead just always wrap the whole query in a 

```sql
SELECT TOP <max rows> * FROM ( ... )
```

before we execute it or something like that

**Comment by flamber** [comment]
Doing a subquery like that would break CTE native questions like this (also see #18685)
```
WITH test_cte AS (
  SELECT 1 as val
)
SELECT * FROM test_cte
```

Another note, `SET ROWCOUNT 0` is currently being used by several to get around #12729

There's no perfect way out of this without breaking something for some people.

**Comment by jeff303** [comment]
I don't think we can leave it totally unconstrained, and it also sounds like a subquery isn't a good idea.

What about simply turning [these](https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/constraints.clj#L5-L11) into settings?  That way admins could adjust the value as needed (perhaps above the default), or maybe even set to `0` to leave unconstrained (along with a big warning flag)?

**Comment by jeff303** [comment]
Solved in 0.43 via #19267.  SQL Server will have a new connection property called _ROWCOUNT Override_.  If left defaulted (i.e. not filled in), then the previous behavior applies (i.e. the 2000 limit).  If this is changed to something (including `0`, which means "unconstrained" in SQL Server's `ROWCOUNT` session variable), then that value will take effect in queries executed against that database.
